Add permissions for launching sailfish-browser and few other applications

spiiroin opened the pull request.

monich commented:
ouch

monich commented on src/jail_run.c:
```diff
@@ -269,6 +269,7 @@ jail_run(
             "setresuid(%u,%u,%u) error: %s", (guint)creds->ruid,
             (guint)creds->euid, (guint)creds->suid, strerror(errno)));
     } else {
+        fflush(0);
```
NULL?

spiiroin replied:
In c literal 0 is automatically promoted to null pointer. Caveat varargs where compiler does not know the intent.

monich replied:
I know, but I'm not a compiler :) Zero is just a number without any meaning attached to it. NULL has additional semantics (hinting that it's a pointer).

spiiroin replied:
NULL it is

monich commented on src/jail_run.c:
```diff
@@ -260,7 +260,7 @@ jail_run(
         g_propagate_error(error, g_error_new(G_UNIX_ERROR, errno,
             "setgroups error: %s", strerror(errno)));
-    } else if (setresgid(creds->rgid, egid, creds->sgid)) {
+    } else if (setresgid(egid, egid, creds->ruid)) {
```
`ruid`? Or did you mean `rgid`?

monich replied:
With or without this change `sailjail cat /proc/self/status` prints the same group ids, I'm not seeing any difference.

spiiroin replied:
yup, gid, not uid. sloppy me ;-)

The difference comes naturally only with privileged apps. Unless I was dreaming / stte -> firejail does not keep egid intact when executing helpers / the final app binary -> using rgid=egid works around that issue. Fixing firejail would be preferable of course, but that would be probably a sizeable task vs. one liner here.

monich replied:
OK, fine. How about adding a comment?

spiiroin replied:
fixed uid -> gid + added a FIXME comment

monich commented on src/jail_rules.c:
```diff
@@ -269,18 +270,18 @@ jail_rules_add_profile(
-
-        path = g_build_filename(conf->perm_dir, file, NULL);
-        g_free(file);
+    char* path = 0;
```
NULL please :)

spiiroin replied:
null it is

monich replied:
yey! :)

monich commented on src/jail_rules.c:
```diff
@@ -269,18 +270,18 @@ jail_rules_add_profile(
+
+    /* Accept only file names with expected suffixes and
+     * no directory components. */
+    if (name && !strchr(name, '/')) {
```
It's a static function, it's only invoked from 2 places and in both places `name` can't be `NULL`. There's no need to check it for `NULL` here.

spiiroin replied:
ack

monich commented on src/jail_rules.c:
```diff
@@ -297,7 +298,7 @@ jail_rules_add_profile(
         g_ptr_array_add(list, jail_rules_profile_new(path, require));
     } else {
-        GWARN("Profile %s doesn't exist", path);
+        GWARN("Profile %s doesn't exist / is not allowed", path ?: name);
```
I would prefer separate warnings for absolute names, invalid suffixes and non-existent files.

spiiroin replied:
I'll shuffle it around a bit, improve diagnostics and relax name vs path portion to better accommodate existence checks before calling this func.

monich commented on src/jail_rules.c:
```diff
@@ -566,6 +573,13 @@ jail_rules_parse_section(
         }
+
+        /* If APPNAME.profile exists, it is implicitly pulled in */
+        {
```
Need to make sure that it exists before calling `jail_rules_add_profile()` to avoid misleading `"XXX doesn't exist"` warnings for something that was never mentioned in the file being parsed.

spiiroin replied:
mm, I guess so. Earlier I had this use jail_rules_add_profile(..., required=false) but the semantics/results from that were not quite as what I expected i.e. this needs to be handled as "required, but only if it exists" kind of way ...

spiiroin replied:
added existence check and tried to make the code more like what there already was

monich commented:
I guess sailjail group in `.desktop` files should be named `[X-Sailjail]` because `desktop-file-install` wants it to start with `X-`

rainemak commented:
Whilst testing saw still:
```
ERROR:test_rules.c:172:test_default_profile: assertion failed (section == "Sailjail"): ("X-Sailjail" == "Sailjail")
Bail out! ERROR:test_rules.c:172:test_default_profile: assertion failed (section == "Sailjail"): ("X-Sailjail" == "Sailjail")
```

spiiroin commented:
Rebased on sailjail 1.0.3. Review commits squashed. Test/devel scripts and associated firejail configs excluded.

monich commented on src/jail_rules_p.h:
```diff
@@ -40,7 +40,7 @@
 #include "jail_rules.h"
 
-#define DEFAULT_PROFILE_SECTION "Sailjail"
+#define DEFAULT_PROFILE_SECTION "X-Sailjail"
```
Make it "X-Sailjail" for .desktop files (which is a special case), "Sailjail" by default.

spiiroin replied:
Dang. Now I notice this.

monich commented:
> Whilst testing saw still:
> 
> ```
> ERROR:test_rules.c:172:test_default_profile: assertion failed (section == "Sailjail"): ("X-Sailjail" == "Sailjail")
> Bail out! ERROR:test_rules.c:172:test_default_profile: assertion failed (section == "Sailjail"): ("X-Sailjail" == "Sailjail")
> ```

That's what unit tests are for!

monich commented:
I was planning to add "X-Sailjail" special case for .desktop files as a separate commit. It's not critical for now, since the section can be specified from the command line - it can be "X-Sailjail", "Y-Sailjail", "Foo", "Bar" or whatever is required in each particular case.

spiiroin commented:
> > Whilst testing saw still:
> > ```
> > ERROR:test_rules.c:172:test_default_profile: assertion failed (section == "Sailjail"): ("X-Sailjail" == "Sailjail")
> > Bail out! ERROR:test_rules.c:172:test_default_profile: assertion failed (section == "Sailjail"): ("X-Sailjail" == "Sailjail")
> > ```
> 
> That's what unit tests are for!

Added commit to fix unit tests.

spiiroin commented:
Rebased on master. Dropped commits dealing with X-Sailjail section as those were handled separately. Cleaned up comment fluff in permission files.

rainemak reviewed (approved):
LGTm
